### PR TITLE
Adding Tailored ControlNet Implementations into Generative Model Application

### LINKED ISSUE #7874: Adding Tailored ControlNet Implementations into Generative Model Application
Adding tailored model implementations into the generative model application to enhance the generation of high-dimensional 3D images.

- [ ] Integrating a tailored ControlNet model into the generative model application to enable the training using high-dimensional 3D images (up to 512 x 512 x 768).

### PR DESCRIPTION
**Pull request by guopengf**
Fixes #7874.

### Description

Integrating a tailored ControlNet model into the generative model application to enable the training using high-dimensional 3D images (up to 512 x 512 x 768).

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

File: monai/apps/generation/maisi/networks/controlnet_maisi.py [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import torch
+
+from typing import Sequence
+from generative.networks.nets.controlnet import ControlNet
```
Comment (KumoLiu) [inline monai/apps/generation/maisi/networks/controlnet_maisi.py:15]: Please follow Dong's PR using `optional_import` here, thank!

File: monai/apps/generation/maisi/networks/controlnet_maisi.py [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+            cross_attention_dim,
+            num_class_embeds,
+            upcast_attention,
+            use_flash_attention,
+            conditioning_embedding_in_channels,
+            conditioning_embedding_num_channels,
+        )
+
+
+    def forward(
```
Comment (KumoLiu) [inline monai/apps/generation/maisi/networks/controlnet_maisi.py:86]: I would suggest splitting the forward method into smaller functional chunks to make it easier to understand and maintain. What do you think?
Something like this: 
```
class CustomControlNet(ControlNet):
    def __init__(self, ...):
        super().__init__(...)

    def forward(self, x, timesteps, controlnet_cond, conditioning_scale=1.0, context=None, class_labels=None):
        emb = self._prepare_time_and_class_embedding(timesteps, class_labels)
        h = self._initial_convolution(x)
        h += self.controlnet_cond_embedding(controlnet_cond)
        h = self._apply_down_blocks(emb, context, h)
        h = self._apply_mid_block(emb, context, h)
        result = self._apply_controlnet_blocks(h, conditioning_scale)
        return result

    def _prepare_time_and_class_embedding(self, timesteps, class_labels):
        ...

    def _initial_convolution(self, x):
        ...

    def _apply_down_blocks(self, emb, context, h):
        ...
        return down_block_res_samples, mid_block_res_sample

    def _apply_mid_block(self, emb, context, h):
        ...

    def _apply_controlnet_blocks(self, h, conditioning_scale):
        ...
        return down_block_res_samples, mid_block_res_sample
```

File: monai/apps/generation/maisi/networks/controlnet_maisi.py (hunk 1 above)
Comment (guopengf) [reply monai/apps/generation/maisi/networks/controlnet_maisi.py]: Thanks! Done

File: monai/apps/generation/maisi/networks/controlnet_maisi.py (hunk 2 above)
Comment (guopengf) [reply monai/apps/generation/maisi/networks/controlnet_maisi.py:97]: Thanks, refactored forward func.

**Comment by guopengf** [comment]
Hi @KumoLiu, would you like to guide me to solve this error? https://github.com/Project-MONAI/MONAI/actions/runs/9670710869/job/26679959381?pr=7875

File: monai/apps/generation/maisi/networks/controlnet_maisi.py [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+        controlnet_down_block_res_samples = ()
+        for down_block_res_sample, controlnet_block in zip(down_block_res_samples, self.controlnet_down_blocks):
+            down_block_res_sample = controlnet_block(down_block_res_sample)
+            controlnet_down_block_res_samples += (down_block_res_sample,)  # type: ignore
+
+        down_block_res_samples = controlnet_down_block_res_samples
+
+        mid_block_res_sample = self.controlnet_mid_block(h)
+
+        return down_block_res_samples, mid_block_res_sample
```
Comment (KumoLiu) [inline monai/apps/generation/maisi/networks/controlnet_maisi.py:180]:
```suggestion
    controlnet_down_block_res_samples = []
    for down_block_res_sample, controlnet_block in zip(down_block_res_samples, self.controlnet_down_blocks):
        down_block_res_sample = controlnet_block(down_block_res_sample)
        controlnet_down_block_res_samples.append(down_block_res_sample)

    mid_block_res_sample = self.controlnet_mid_block(h)

    return controlnet_down_block_res_samples, mid_block_res_sample
```

File: tests/test_controlnet_maisi.py [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+
+from monai.networks import eval_mode
+from monai.utils import optional_import
+
+_, has_generative = optional_import("generative")
+
+if has_generative:
+    from monai.apps.generation.maisi.networks.controlnet_maisi import ControlNetMaisi
+
+TEST_CASES = [
```
Comment (KumoLiu) [inline tests/test_controlnet_maisi.py:28]: Thanks for the quick update! Overall loos good to me. Could you please add more test cases to include more variations of parameter combinations and also for error catching?

**Comment by KumoLiu** [comment]
Hi @ericspod @Nic-Ma @mingxin-zheng, if you don't have more comments on this one, I will try to trigger more tests and merge this one. Thanks!

**Comment by KumoLiu** [comment]
Hi @guopengf, could you please help solve the conflict then I will try to merge this PR, thanks!

**Comment by guopengf** [comment]
> Hi @guopengf, could you please help solve the conflict then I will try to merge this PR, thanks!

Done, thanks!

**Review by KumoLiu — APPROVED** [review]

**Review by ericspod — APPROVED** [review]
I'm good here!

**Comment by KumoLiu** [comment]
Add `monai-generative`, will cause this error:
```

[2024-06-29T13:15:38.518Z] ======================================================================

[2024-06-29T13:15:38.518Z] FAIL: test_visible_devices (tests.test_set_visible_devices.TestVisibleDevices)

[2024-06-29T13:15:38.518Z] ----------------------------------------------------------------------

[2024-06-29T13:15:38.518Z] Traceback (most recent call last):

[2024-06-29T13:15:38.518Z]   File "/home/jenkins/agent/workspace/MONAI-premerge/monai/tests/test_set_visible_devices.py", line 38, in test_visible_devices

[2024-06-29T13:15:38.518Z]     self.assertEqual(num_gpus_before, num_gpus_after)

[2024-06-29T13:15:38.518Z] AssertionError: 0 != 1

[2024-06-29T13:15:38.518Z] 

[2024-06-29T13:15:38.518Z] ----------------------------------------------------------------------
```

Do you have any insights on this one? @ericspod @Nic-Ma @mingxin-zheng

File: monai/apps/generation/maisi/networks/controlnet_maisi.py [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+        return h
+
+    def _apply_down_blocks(self, emb, context, h):
+        # 4. down
+        if context is not None and self.with_conditioning is False:
+            raise ValueError("model should have with_conditioning = True if context is provided")
+        down_block_res_samples: list[torch.Tensor] = [h]
+        for downsample_block in self.down_blocks:
+            h, res_samples = downsample_block(hidden_states=h, temb=emb, context=context)
+            for residual in res_samples:
```
Comment (mingxin-zheng) [inline monai/apps/generation/maisi/networks/controlnet_maisi.py:159]: Hi @guopengf , would the use of `extend` be okay to replace iterating append?
```
down_block_res_samples.extend(res_samples)
```

**Comment by mingxin-zheng** [comment]
> Do you have any insights on this one? @ericspod @Nic-Ma @mingxin-zheng

It seems the error is caused by this line:
```python
ControlNet, has_controlnet = optional_import("generative.networks.nets.controlnet", name="ControlNet")
```

**Comment by mingxin-zheng** [comment]
Update: pinpointed this class is causing the issue for unknown reason @KumoLiu 

https://github.com/Project-MONAI/GenerativeModels/blob/a9b17d461d42b7b3487d0e29be265d9373b11d67/generative/networks/layers/vector_quantizer.py#L20

**Review by mingxin-zheng — APPROVED** [review]
Looks good to me.

**Comment by mingxin-zheng** [comment]
I am merging this to unblock https://github.com/Project-MONAI/MONAI/pull/7861.

**Pull request merged**